docs/src/proposals/epoch-boundary-stake-reward.md

```diff
@@ -0,0 +1,26 @@
+---
+title: Stake Reward At Epoch Boundary
+---
+
+## Problem
+
+With the increase of number of stake accounts, computing and redeeming the stake rewards at the start slot of the epoch boundary becomes very expensive. Currently, with 550K stake accounts, the stake reward time has already taken more than 10 seconds. This prolonged computation slows down the network, and can cause large number of forks at the epoch boundary, which make the matter even worse.
+
+## Proposed Solutions
+
+Instead of computing and reward stake accounts at epoch boundary, we will decouple reward computation and reward credit for stake accounts. At the start of the new epoch, a background thread is forked to compute the rewards. At slot `N` after the epoch, the rewards will be credited to all stake accounts. We could call these `epoch_start..epoch_start+N` slots are `rewarding interval`.  `N` is chosen to be sufficiently large so that the background computation should have completed and the result of the reward computation is available at the end of `rewarding interval`. `N` can be fixed such as 100 (roughly equivalent to 50 seconds), or chosen as a function of the number of stake accounts, `f(num_stake_accounts)`. When reached slot `N`, the bank will fetch the reward computation results from the background thread, and credit the rewards to all the stake accounts.
```

Comment:
what is the breakdown of calculation time vs
store time for 500k writes + whatever bank freeze parts run over the 500k(?) writes in the slot?

Reply:
```
> column -t -s, a.txt
                                    us
thread_pool_creation_us             3023i
apply_feature_activations           5937i
activate_epoch_us                   153867i
update_epoch_stakes_us              9695i
update_rewards_with_thread_pool_us  11448012i
load_vote_and_stake_accounts_us     7560376i
calculate_points_us                 59366i
redeem_rewards_us                   938866i
tore_stake_accounts_us              2571345i
store_vote_accounts_us              39301i
```

Reply:
A breakdown for the reward computation time shown above. 
Storing 500K account will take about 2.5s. 
We have discussed another idea to sharding the store over multiple slots, i.e. storing 50K account per slot over 10 slots. The concern is that it maybe unfair for some accounts to get the rewards credited earlier than other accounts.

Reply:
> 2.5s
We can 'store' things into the write cache in a separate area ahead of time, invisibly, and 'move' the stored accounts into the actual write cache in a few ways. Updating the index will be tricky. An issue will be if we skip slot N and have to do this multiple times. 2.5s to store seems likely to almost guarantee we will skip slot N usually, possibly. That's how it works at beginning of epoch today, right? We could update the index and have a write cache ready for an optimistic slot N. If it fails on that slot, then we'll have to do more work and undo some of what was expected to happen.

Reply:
> > The concern is that it maybe unfair for some accounts to get the rewards credited earlier than other accounts.
an idea is to sort accounts by pubkey. Store 1/10 of them, starting at an offset into the sorted list, doing 1/10 per slot. The offset could be a deterministic calculation based off the first byte after hashing the total rewards or the middle pubkey or the epoch + middle pubkey or hashing the pubkey of the leader on slot N, or... Or some other scheme that allowed us to randomly distribute the rewards over 10 slots so the pain is random. We basically say, rewards distribution starts at N and ends at N+10, and each reward will be paid by the end. And people can pre-calculate their position if they want. Just brainstorming. 2.5s still seems bad.

Reply:
yeah. 2.5s is bad. And credit all accounts at one slot is not going to scale. 
I think we can relax the strict constraint to credit the account at the same slots. Like your proposal, we can spread the store over 10 slots, but block the access for those account during the additional 10 slots.
I will update the proposal to include this idea.

Reply:
> The concern is that it maybe unfair for some accounts to get the rewards credited earlier than other accounts.

I don't think this is a reasonable concern assuming we lock _all_ stake accounts for the duration of the rewards payout and write back.  Everyone takes the same hit once during the first epoch where we enable this, then everyone is on a one epoch cadence with the same delay from the epoch start for the remainder of history

Reply:
Yes.